securing_apps/topics/cli-apps/cli-apps.adoc

```diff
@@ -0,0 +1,137 @@
+[[_kcinit]]
+
+== Securing Command Line Applications
+
+Many applications need a command line interface to administer and configure themselves.
+If these command line interfaces are invoking on back-end admin REST interfaces that
+are secured by {project_name}, they need a way to obtain a token to invoke on them.
+This chapter discusses some utilities that {project_name} has that you can
+use to secure your command line interfaces.
+
+The initial and standard way to obtain a token for a command line application
+is to prompt the user for username and password and invoke on the <<_resource_owner_password_credentials_flow, Resource Owners Credential Grant>>
+ OAuth endpoint that {project_name} has.  While a completely acceptable solution,
+there are some disadvantages to this approach.  The obvious one is that you'll have to
+code the protocol and manage the tokens.  This means your CLI app will have to manage
+and be aware of token timeouts and refresh tokens.  The second disadvantage is
+that if there are other CLI apps using {project_name}, you'll have to relogin
+to each different CLI app that is secured by {project_name}.  No SSO on the command line.
+Finally, your app will have to be hard coded to specific credential types.  If you've
+only implemented password support and the {project_name} realm admin decides that an
+OTP will also be required to login, you will have to change your CLI application to
+support it.
+
+=== Kcinit CLI Utility
+
+To help developers solve some of the CLI security issues, the {project_name} team created the `kcinit` command line utility.
+This utility handles login with a Keycloak realm and stores and manages tokens locally.
+These tokens can then be passed along to your CLI applications as a command line switch or an environment variable.
+Using this tool is best described with a real example.  {project_name} has integration with the Kubernetes project.  Kubernetes
+has a backend REST API that is secured with bearer tokens and accessed with the command line utility `kubectl`.
+You can use `kcinit` to obtain and manage tokens and pass these tokens along via the `kubectl` switch `--token`.
+For example:
+
+----
+     kubectl --token=$(kcinit token)
+----
+
+`kcinit` would prompt you for login using a text-based interface.  `kcinit` would perform the OAuth protocol behind the
+scenes to obtain a token, and output the token to STDOUT.  Any access or refresh token that `kcinit` obtains
+is managed and stored locally.  This allows you to login once, and `kcinit` will handle whether you have to log
+in again.  It handles refreshing tokens automatically.  If you are using `kcinit` as command line SSO with other CLI apps,
+`kcinit` can also perform token exchanges to obtain the correct token for the exact REST backend your CLI app is invoking on.
```

Comment:
Changing this to active voice with a non-conditional verb phrasing...

After entering this example, `kcinit` prompts you to login using a text-based interface. Then, `kcinit` performs the OAuth protocol behind the scenes to obtain a token, and outputs the token to STDOUT.